prov/shm/src/smr_dsa.c

```diff
@@ -97,6 +98,40 @@ struct smr_dsa_context {
 	unsigned long page_fault_stats[2];
 };
 
+struct dsa_ops {
+	struct accfg_device *(*accfg_wq_get_device)(struct accfg_wq *wq);
+	int (*accfg_device_get_cdev_major)(struct accfg_device *dev);
+	int (*accfg_wq_get_cdev_minor)(struct accfg_wq *wq);
+	int (*accfg_new)(struct accfg_ctx **ctx);
+	enum accfg_device_state (*accfg_device_get_state)(struct accfg_device *device);
+	uint64_t (*accfg_device_get_gen_cap)(struct accfg_device *device);
+	int (*accfg_device_get_numa_node)(struct accfg_device *device);
+	enum accfg_wq_state (*accfg_wq_get_state)(struct accfg_wq *wq);
+	uint64_t (*accfg_wq_get_max_transfer_size)(struct accfg_wq *wq);
+	enum accfg_wq_type (*accfg_wq_get_type)(struct accfg_wq *wq);
+	enum accfg_wq_mode (*accfg_wq_get_mode)(struct accfg_wq *wq);
+	const char *(*accfg_wq_get_devname)(struct accfg_wq *wq);
+	struct accfg_ctx *(*accfg_unref)(struct accfg_ctx *ctx);
+
+	struct accfg_device *(*accfg_device_get_first)(struct accfg_ctx *ctx);
+	struct accfg_device *(*accfg_device_get_next)(struct accfg_device *device);
+	struct accfg_wq *(*accfg_wq_get_first)(struct accfg_device *device);
+	struct accfg_wq *(*accfg_wq_get_next)(struct accfg_wq *wq);
+};
+
+#define device_foreach(ctx, device) \
+	for (device = dsa_ops.accfg_device_get_first(ctx); \
+	     device != NULL; \
+	     device = dsa_ops.accfg_device_get_next(device))
+
+
+#define wq_foreach(device, wq) \
+	for (wq = dsa_ops.accfg_wq_get_first(device); \
+	     wq != NULL; \
+	     wq = dsa_ops.accfg_wq_get_next(wq))
```

Comment:
Consider adding 'dsa' somewhere in the above macro names to avoid confusion and namespace collision. E.g. dsa_foreach_device / dsa_foreach_wq

Reply:
good call